### ISSUE TITLE
hostcheckstop refactor

### ISSUE THREAD

**Issue by geissonator** [openbmc/openbmc#766]
Look at skeleton/hostcheckstop/host_checkstop_obj.c
- Define yaml (if any interfaces or properties required)
- Compile down with sdbusplus
- Write in c++ and use sdbusplus for dbus interactions

**Comment by mine260309** [comment]
The inteface `org.openbmc.control.Checkstop` seems to have no methods nor properties.
Is it meaningful to define an empty interface in `phosphor-dbus-interfaces`?

**Comment by geissonator** [comment]
Discussed a bit in design meeting.  The thought is to not have a sensor or sdbus property for this.  We'll simply have a process that monitors the checkstop gpio, and issues a systemd transition request to reboot system when it hits.  Patrick proposed we rename this to something more generic like - "Host Termination Signal".  We may also need to create an elog when this happens.

**Comment by spinler** [comment]
For checkstop debug we'll need a way to disable this.  Is it as easy as stopping the service?

**Comment by geissonator** [comment]
Good point, yes, that should be a requirement here.

**Comment by geissonator** [comment]
And also, probably shouldn't use systemd directly, should use the new Host interface to restart things (it calls systemd).

**Comment by geissonator** [comment]
This will use the new quiesce target functions being created in #1033

**Comment by vishwabmc** [comment]
Eddie is currently doing some GPIO support and I feel we can park this story for future sprints considering that its needed for phase-5 and gpio stuff is being updated.

@geissonator can we move this out ?

**Comment by vishwabmc** [comment]
There was a design discussion on this and here is the key takeaway.

* 4.9 Kernel is coming with a IOCTL for GPIO which gives the FD for a passed in GPIO line.
* We can still go ahead and do the work now without waiting for 4.9
  - idea is to have a stub that will do a open on some existing filesystem entity ( like a fifo, pipe, file etc ).
  - and let that give the FD and when 4.9 is in, this would be augmented to using that GPIO IOCTL
* The application will be passed the gpio chip number and the gpio number so that the app does not do any
   interpretation.
* then the usual sd_event mechanism can be used to wait on the gpio assertions

* The application does not need to worry about setting the GPIO direction as its a IN pin.

**Comment by vishwabmc** [comment]
- [ ]

**Comment by geissonator** [comment]
Lets touch base in Monday call on what will happen when we hit the gpio.

**Comment by sivassrr** [comment]
https://gerrit.openbmc-project.xyz/#/c/2938/ - Do we need to handle anything else in our test. 
(Hope this changes cover for our test, let know if we need to add any).  We can add new issue.

**Comment by vishwabmc** [comment]
in yesterday's meeting, one of the idea was to use /dev/input/event* to monitor the GPIO line.

I looked at Palmetto / Barreleye / Witherspoon system and did not see /dev/input.

Looks like we don't have input device support in the Kernel ?

@bradbishop ,  looks like its wise to wait till we move to 4.10 and use the IOCTL ?.

@geissonator I see its marked for `Phase-5` and I believe this needs to be taken up when we are ready with the support ?

@shenki you have any thoughts ?

**Comment by vishwabmc** [comment]
@rfrandse I see this issue has history of its ready for 17.02 etc and I feel its not appropriate since we don't even seem to have the needed support ?

**Comment by geissonator** [comment]
This is needed for phase 4 so we can put the error creation code in (so external users are notified when we hit a checkstop).  #1266 is the story for that.

**Comment by vishwabmc** [comment]
Needs https://github.com/openbmc/openbmc/issues/1387

**Comment by shenki** [comment]
The suggestion relating to input devices was to use the GPIO keys driver:

https://www.kernel.org/doc/Documentation/devicetree/bindings/input/gpio-keys.txt

In addition entries to the device tree, you would need to enable `CONFIG_KEYBOARD_GPIO` in the kernel.

**Comment by vishwabmc** [comment]
I talked to @bradbishop yesterday and he was saying that relaying the GPIO values onto /dev/input/ needs some work there. We would be discussing on that in today's design meeting.

@shenki if you have any thought, please let me know.
